File: sdks/kotlin/demo/quickstart.kt
// nRouter Kotlin Example
//
// Demonstrates:
// 1. Basic chat completion with Coroutines
// 2. Response metadata inspection (cost, tokens, request ID, cache)
// 3. Typed error handling (NRouterError)
// 4. Prompt variables and custom configuration
//
// Maven: ai.nrouter:nrouter-sdk-kotlin:3.1.2

package ai.nrouter.examples

import ai.nrouter.sdk.NRouter
import ai.nrouter.sdk.NRouterError
import kotlinx.coroutines.runBlocking
import org.json.JSONArray
import org.json.JSONObject

fun main() = runBlocking {
    val apiKey = System.getenv("NROUTER_API_KEY")
    if (apiKey == null) {
        println("Please set NROUTER_API_KEY environment variable.")
        return@runBlocking
    }

    try {
        // ━━━ 1. Client Initialization ━━━━━━━━━━━━━━━━━━━━━━━━
        val client = NRouter(apiKey = apiKey)

        // ━━━ 2. Chat Completion ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
        println("--- Sending Chat Request ---")
        val requestBody = JSONObject()
            .put("model", "gpt-5.4-mini")
            .put("messages", JSONArray().apply {
                put(JSONObject().put("role", "system").put("content", "You are a helpful assistant."))
                put(JSONObject().put("role", "user").put("content", "Explain what an LLM Gateway is in two sentences."))
            })
            .put("max_tokens", 100)

        val response = client.chatCompletions(requestBody)

        val content = response.body
            .getJSONArray("choices")
            .getJSONObject(0)
            .getJSONObject("message")
            .getString("content")

        println("\nResponse:\n$content")

        // ━━━ 3. Response Metadata Inspection ━━━━━━━━━━━━━━━━
        val meta = response.meta
        println("\n--- Response Metadata ---")
        println("Request ID:    ${meta.requestID ?: "n/a"}")
        println("Model Served:  ${meta.model ?: "n/a"}")
        println("Tokens:        Input: ${meta.inputTokens ?: 0}, Output: ${meta.outputTokens ?: 0}, Total: ${meta.totalTokens ?: 0}")

        if (meta.isPriced && meta.cost != null) {
            println("Exact Cost:    $${meta.cost}")
        } else {
            println("Cost Status:   ${meta.costStatus ?: "unpriced"}")
        }

        meta.limitSource?.let { println("Limit Source:  $it") }

    } catch (e: NRouterError.GuardrailBlocked) {
        println("\nBlocked by Guardrail: ${e.body.message}")
    } catch (e: NRouterError.Credit) {
        println("\nCredit Shortfall: ${e.body.message}")
    } catch (e: NRouterError.RateLimit) {
        println("\nRate Limit Hit from ${e.body.limitSource ?: "gateway"}: ${e.body.message}")
        if (e.isRetryable) println("Retryable with exponential backoff.")
    } catch (e: NRouterError.Authentication) {
        println("\nInvalid API Key: ${e.body.message}")
    } catch (e: NRouterError.Transport) {
        println("\nTransport / Network failure: ${e.message}")
    } catch (e: NRouterError) {
        println("\nGateway error [${e.code ?: "code"}]: ${e.body.message}")
    } catch (e: Exception) {
        println("\nUnexpected error: ${e.message}")
    }
}

File: sdks/swift/demo/quickstart.swift
// nRouter Swift Example
//
// Demonstrates:
// 1. Basic chat completion. The model must be one whose provider serves
//    /v1/chat/completions — Anthropic serves /v1/messages only.
// 2. Response metadata inspection (cost, tokens, request ID, cache)
// 3. Error handling with typed NRouterError
// 4. Custom base URL & configuration
//
// Swift Package Manager:
// .package(url: "https://github.com/nRouterGateway/nrouter-sdk.git", from: "3.1.2")

import Foundation
import NRouter

@main
struct NRouterExample {
    static func main() async {
        guard let apiKey = ProcessInfo.processInfo.environment["NROUTER_API_KEY"] else {
            print("Please set NROUTER_API_KEY environment variable.")
            return
        }

        do {
            // ━━━ 1. Client Initialization ━━━━━━━━━━━━━━━━━━━━━━━━
            let client = try NRouter(apiKey: apiKey)

            // ━━━ 2. Chat Completion ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
            print("--- Sending Chat Request ---")
            let requestBody: [String: Any] = [
                "model": "gpt-5.4-mini",
                "messages": [
                    ["role": "system", "content": "You are a concise AI assistant."],
                    ["role": "user", "content": "Explain what an LLM Gateway is in two sentences."]
                ],
                "max_tokens": 100
            ]

            let response = try await client.chatCompletions(requestBody)

            if let choices = response.body["choices"] as? [[String: Any]],
               let firstChoice = choices.first,
               let message = firstChoice["message"] as? [String: Any],
               let content = message["content"] as? String {
                print("\nResponse:")
                print(content)
            }

            // ━━━ 3. Response Metadata Inspection ━━━━━━━━━━━━━━━━
            let meta = response.meta
            print("\n--- Response Metadata ---")
            print("Request ID:    \(meta.requestID ?? "n/a")")
            print("Model Served:  \(meta.model ?? "n/a")")
            print("Tokens:        Input: \(meta.inputTokens ?? 0), Output: \(meta.outputTokens ?? 0), Total: \(meta.totalTokens ?? 0)")

            if meta.isPriced, let cost = meta.cost {
                print("Exact Cost:    $\(cost)")
            } else {
                print("Cost Status:   \(meta.costStatus ?? "unpriced")")
            }

            if let limitSource = meta.limitSource {
                print("Limit Source:  \(limitSource)")
            }

        } catch let error as NRouterError {
            // ━━━ 4. Typed Error Handling ━━━━━━━━━━━━━━━━━━━━━━━━
            print("\n[nRouter Error]: \(error.localizedDescription)")
            switch error {
            case .guardrailBlocked(let body):
                print("Blocked by Guardrail: \(body.message)")
            case .credit(let body):
                print("Credit Shortfall: \(body.message)")
            case .rateLimit(let body):
                print("Rate Limit Hit from \(body.limitSource ?? "gateway"): \(body.message)")
                if error.isRetryable {
                    print("This error is retryable with backoff.")
                }
            case .authentication(let body):
                print("Invalid API Key: \(body.message)")
            case .notFound(let body):
                print("Model Not Found: \(body.message)")
            case .transport(let message):
                print("Transport / Network failure: \(message)")
            default:
                print("Gateway error code: \(error.code ?? "unknown")")
            }
        } catch {
            print("Unexpected error: \(error)")
        }
    }
}
